File: sdk/rust/src/account/account_info.rs
/*
 * ‌
 * Hedera Rust SDK
 * ​
 * Copyright (C) 2022 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

use hedera_proto::services;
use prost::Message;
use time::{
    Duration,
    OffsetDateTime,
};

use crate::protobuf::ToProtobuf;
use crate::{
    AccountId,
    FromProtobuf,
    Hbar,
    Key,
    LedgerId,
    PublicKey,
    StakingInfo,
    Tinybar,
};

/// Response from [`AccountInfoQuery`][crate::AccountInfoQuery].
#[derive(Debug, Clone)]
#[cfg_attr(feature = "ffi", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "ffi", serde(rename_all = "camelCase"))]
pub struct AccountInfo {
    /// The account that is being referenced.
    pub account_id: AccountId,

    /// The Contract Account ID comprising of both the contract instance and the cryptocurrency
    /// account owned by the contract instance, in the format used by Solidity.
    pub contract_account_id: String,

    /// If true, then this account has been deleted, it will disappear when it expires, and all
    /// transactions for it will fail except the transaction to extend its expiration date.
    pub is_deleted: bool,

    /// The Account ID of the account to which this is proxy staked.
    ///
    /// If `proxy_account_id` is `None`, an invalid account, or an account that isn't a node,
    /// then this account is automatically proxy staked to a node chosen by the network,
    /// but without earning payments.
    ///
    /// If the `proxy_account_id` account refuses to accept proxy staking, or if it is not currently
    /// running a node, then it will behave as if `proxy_account_id` is `None`.
    #[deprecated]
    pub proxy_account_id: Option<AccountId>,

    /// The total number of hbars proxy staked to this account.
    pub proxy_received: Hbar,

    /// The key for the account, which must sign in order to transfer out, or to modify the
    /// account in any way other than extending its expiration date.
    pub key: Key,

    /// Current balance of the referenced account.
    pub balance: Hbar,

    /// The threshold amount for which an account record is created (and this account
    /// charged for them) for any send/withdraw transaction.
    #[deprecated]
    pub send_record_threshold: Hbar,

    /// The threshold amount for which an account record is created
    /// (and this account charged for them) for any transaction above this amount.
    #[deprecated]
    pub receive_record_threshold: Hbar,

    /// If true, no transaction can transfer to this account unless signed by
    /// this account's key.
    pub is_receiver_signature_required: bool,

    /// The time at which this account is set to expire.
    #[cfg_attr(
        feature = "ffi",
        serde(with = "serde_with::As::<Option<serde_with::TimestampNanoSeconds>>")
    )]
    pub expiration_time: Option<OffsetDateTime>,

    /// The duration for expiration time will extend every this many seconds.
    #[cfg_attr(
        feature = "ffi",
        serde(with = "serde_with::As::<Option<serde_with::DurationSeconds<i64>>>")
    )]
    pub auto_renew_period: Option<Duration>,

    /// The memo associated with the account.
    pub account_memo: String,

    /// The number of NFTs owned by this account
    pub owned_nfts: u64,

    /// The maximum number of tokens that an Account can be implicitly associated with.
    pub max_automatic_token_associations: u32,

    /// The alias of this account.
    pub alias_key: Option<PublicKey>,

    /// The ethereum transaction nonce associated with this account.
    pub ethereum_nonce: u64,

    /// The ledger ID the response was returned from.
    pub ledger_id: LedgerId,

    /// Staking metadata for this account.
    pub staking: Option<StakingInfo>,
}

impl AccountInfo {
    /// Create a new `AccountInfo` from protobuf-encoded `bytes`.
    ///
    /// # Errors
    /// - [`Error::FromProtobuf`](crate::Error::FromProtobuf) if decoding the bytes fails to produce a valid protobuf.
    /// - [`Error::FromProtobuf`](crate::Error::FromProtobuf) if decoding the protobuf fails.
    pub fn from_bytes(bytes: &[u8]) -> crate::Result<Self> {
        FromProtobuf::<services::crypto_get_info_response::AccountInfo>::from_bytes(bytes)
    }

    /// Convert `self` to a protobuf-encoded [`Vec<u8>`].
    #[must_use]
    pub fn to_bytes(&self) -> Vec<u8> {
        #[allow(deprecated)]
        services::crypto_get_info_response::AccountInfo {
            account_id: Some(self.account_id.to_protobuf()),
            contract_account_id: self.contract_account_id.clone(),
            deleted: self.is_deleted,
            proxy_received: self.proxy_received.to_tinybars(),
            key: Some(self.key.to_protobuf()),
            balance: self.balance.to_tinybars() as u64,
            receiver_sig_required: self.is_receiver_signature_required,
            expiration_time: self.expiration_time.to_protobuf(),
            auto_renew_period: self.auto_renew_period.to_protobuf(),
            memo: self.account_memo.clone(),
            owned_nfts: self.owned_nfts as i64,
            max_automatic_token_associations: self.max_automatic_token_associations as i32,
            alias: self.alias_key.as_ref().map(ToProtobuf::to_bytes).unwrap_or_default(),
            ledger_id: self.ledger_id.to_bytes(),
            ethereum_nonce: self.ethereum_nonce as i64,
            staking_info: self.staking.to_protobuf(),

            // implemented deprecated fields
            proxy_account_id: self.proxy_account_id.to_protobuf(),
            generate_receive_record_threshold: self.receive_record_threshold.to_tinybars() as u64,
            generate_send_record_threshold: self.send_record_threshold.to_tinybars() as u64,

            // unimplemented fields
            live_hashes: Vec::default(),
            token_relationships: Vec::default(),

            // unimplemented deprecated fields
            ..Default::default()
        }
        .encode_to_vec()
    }
}

impl FromProtobuf<services::response::Response> for AccountInfo {
    fn from_protobuf(pb: services::response::Response) -> crate::Result<Self>
    where
        Self: Sized,
    {
        let response = pb_getv!(pb, CryptoGetInfo, services::response::Response);
        let info = pb_getf!(response, account_info)?;
        Self::from_protobuf(info)
    }
}

impl FromProtobuf<services::crypto_get_info_response::AccountInfo> for AccountInfo {
    fn from_protobuf(pb: services::crypto_get_info_response::AccountInfo) -> crate::Result<Self>
    where
        Self: Sized,
    {
        let key = pb_getf!(pb, key)?;
        let account_id = pb_getf!(pb, account_id)?;
        let alias_key = PublicKey::from_alias_bytes(&pb.alias)?;
        let ledger_id = LedgerId::from_bytes(pb.ledger_id);
        let staking = Option::from_protobuf(pb.staking_info)?;

        #[allow(deprecated)]
        Ok(Self {
            ledger_id,
            staking,
            account_id: AccountId::from_protobuf(account_id)?,
            contract_account_id: pb.contract_account_id,
            is_deleted: pb.deleted,
            proxy_received: Hbar::from_tinybars(pb.proxy_received),
            key: Key::from_protobuf(key)?,
            balance: Hbar::from_tinybars(pb.balance as Tinybar),
            expiration_time: pb.expiration_time.map(Into::into),
            auto_renew_period: pb.auto_renew_period.map(Into::into),
            account_memo: pb.memo,
            owned_nfts: pb.owned_nfts as u64,
            max_automatic_token_associations: pb.max_automatic_token_associations as u32,
            alias_key,
            ethereum_nonce: pb.ethereum_nonce as u64,
            is_receiver_signature_required: pb.receiver_sig_required,

            // deprecated fields
            proxy_account_id: Option::from_protobuf(pb.proxy_account_id)?,
            send_record_threshold: Hbar::from_tinybars(pb.generate_send_record_threshold as i64),
            receive_record_threshold: Hbar::from_tinybars(
                pb.generate_receive_record_threshold as i64,
            ),
        })
    }
}

File: sdk/swift/Sources/Hedera/Account/AccountInfo.swift
/*
 * ‌
 * Hedera Swift SDK
 * ​
 * Copyright (C) 2022 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import Foundation
import HederaProtobufs

/// Response from `AccountInfoQuery`.
public final class AccountInfo: Codable {
    internal init(
        accountId: AccountId,
        contractAccountId: String,
        isDeleted: Bool,
        proxyAccountId: AccountId?,
        proxyReceived: Hbar,
        key: Key,
        balance: Hbar,
        sendRecordThreshold: Hbar,
        receiveRecordThreshold: Hbar,
        isReceiverSignatureRequired: Bool,
        expirationTime: Timestamp?,
        autoRenewPeriod: Duration?,
        accountMemo: String,
        ownedNfts: UInt64,
        maxAutomaticTokenAssociations: UInt32,
        aliasKey: PublicKey?,
        ethereumNonce: UInt64,
        ledgerId: LedgerId,
        staking: StakingInfo?
    ) {
        self.accountId = accountId
        self.contractAccountId = contractAccountId
        self.isDeleted = isDeleted
        self.proxyAccountId = proxyAccountId
        self.proxyReceived = proxyReceived
        self.key = key
        self.balance = balance
        self.sendRecordThreshold = sendRecordThreshold
        self.receiveRecordThreshold = receiveRecordThreshold
        self.isReceiverSignatureRequired = isReceiverSignatureRequired
        self.expirationTime = expirationTime
        self.autoRenewPeriod = autoRenewPeriod
        self.accountMemo = accountMemo
        self.ownedNfts = ownedNfts
        self.maxAutomaticTokenAssociations = maxAutomaticTokenAssociations
        self.aliasKey = aliasKey
        self.ethereumNonce = ethereumNonce
        self.ledgerId = ledgerId
        self.staking = staking
    }

    /// The account that is being referenced.
    public let accountId: AccountId

    /// The Contract Account ID comprising of both the contract instance and the cryptocurrency
    /// account owned by the contract instance, in the format used by Solidity.
    public let contractAccountId: String

    /// If true, then this account has been deleted, it will disappear when it expires, and all
    /// transactions for it will fail except the transaction to extend its expiration date.
    public let isDeleted: Bool

    /// The Account ID of the account to which this is proxy staked.
    ///
    /// If `proxy_account_id` is `None`, an invalid account, or an account that isn't a node,
    /// then this account is automatically proxy staked to a node chosen by the network,
    /// but without earning payments.
    ///
    /// If the `proxy_account_id` account refuses to accept proxy staking, or if it is not currently
    /// running a node, then it will behave as if `proxy_account_id` is `None`.
    // @available(*, deprecated)
    public let proxyAccountId: AccountId?

    /// The total number of HBARs proxy staked to this account.
    public let proxyReceived: Hbar

    /// The key for the account, which must sign in order to transfer out, or to modify the
    /// account in any way other than extending its expiration date.
    public let key: Key

    /// Current balance of the referenced account.
    public let balance: Hbar

    /// The threshold amount for which an account record is created (and this account
    /// charged for them) for any send/withdraw transaction.
    // @available(*, deprecated)
    public let sendRecordThreshold: Hbar

    /// The threshold amount for which an account record is created
    /// (and this account charged for them) for any transaction above this amount.
    // @available(*, deprecated)
    public let receiveRecordThreshold: Hbar

    /// If true, no transaction can transfer to this account unless signed by
    /// this account's key.
    public let isReceiverSignatureRequired: Bool

    /// The TimeStamp time at which this account is set to expire.
    public let expirationTime: Timestamp?

    /// The duration for expiration time will extend every this many seconds.
    public let autoRenewPeriod: Duration?

    /// The memo associated with the account.
    public let accountMemo: String

    /// The number of NFTs owned by this account
    public let ownedNfts: UInt64

    /// The maximum number of tokens that an Account can be implicitly associated with.
    public let maxAutomaticTokenAssociations: UInt32

    /// The public key which aliases to this account.
    public let aliasKey: PublicKey?

    /// The ethereum transaction nonce associated with this account.
    public let ethereumNonce: UInt64

    /// The ledger ID the response was returned from.
    public let ledgerId: LedgerId

    /// Staking metadata for this account.
    public let staking: StakingInfo?

    public static func fromBytes(_ bytes: Data) throws -> Self {
        try Self(fromProtobufBytes: bytes)
    }

    public func toBytes() -> Data {
        self.toProtobufBytes()
    }
}

extension AccountInfo: TryProtobufCodable {
    internal typealias Protobuf = Proto_CryptoGetInfoResponse.AccountInfo

    internal convenience init(fromProtobuf proto: Protobuf) throws {
        let expirationTime = proto.hasExpirationTime ? proto.expirationTime : nil
        let autoRenewPeriod = proto.hasAutoRenewPeriod ? proto.autoRenewPeriod : nil
        let staking = proto.hasStakingInfo ? proto.stakingInfo : nil
        let proxyAccountId = proto.hasProxyAccountID ? proto.proxyAccountID : nil

        self.init(
            accountId: try .fromProtobuf(proto.accountID),
            contractAccountId: proto.contractAccountID,
            isDeleted: proto.deleted,
            proxyAccountId: try .fromProtobuf(proxyAccountId),
            proxyReceived: Hbar.fromTinybars(proto.proxyReceived),
            key: try .fromProtobuf(proto.key),
            balance: .fromTinybars(Int64(proto.balance)),
            sendRecordThreshold: Hbar.fromTinybars(Int64(proto.generateSendRecordThreshold)),
            receiveRecordThreshold: Hbar.fromTinybars(Int64(proto.generateReceiveRecordThreshold)),
            isReceiverSignatureRequired: proto.receiverSigRequired,
            expirationTime: .fromProtobuf(expirationTime),
            autoRenewPeriod: .fromProtobuf(autoRenewPeriod),
            accountMemo: proto.memo,
            ownedNfts: UInt64(proto.ownedNfts),
            maxAutomaticTokenAssociations: UInt32(proto.maxAutomaticTokenAssociations),
            aliasKey: try .fromAliasBytes(proto.alias),
            ethereumNonce: UInt64(proto.ethereumNonce),
            ledgerId: .fromBytes(proto.ledgerID),
            staking: try .fromProtobuf(staking)
        )
    }

    internal func toProtobuf() -> Protobuf {
        .with { proto in
            proto.accountID = accountId.toProtobuf()
            proto.contractAccountID = contractAccountId
            proto.deleted = isDeleted
            proxyAccountId?.toProtobufInto(&proto.proxyAccountID)
            proto.proxyReceived = proxyReceived.toTinybars()
            proto.key = key.toProtobuf()
            proto.balance = UInt64(balance.toTinybars())
            proto.generateSendRecordThreshold = UInt64(sendRecordThreshold.toTinybars())
            proto.generateReceiveRecordThreshold = UInt64(receiveRecordThreshold.toTinybars())
            proto.receiverSigRequired = isReceiverSignatureRequired
            expirationTime?.toProtobufInto(&proto.expirationTime)
            autoRenewPeriod?.toProtobufInto(&proto.autoRenewPeriod)
            proto.memo = accountMemo
            proto.ownedNfts = Int64(ownedNfts)
            proto.maxAutomaticTokenAssociations = Int32(maxAutomaticTokenAssociations)
            proto.alias = aliasKey?.toProtobufBytes() ?? Data()
            proto.ethereumNonce = Int64(ethereumNonce)
            proto.ledgerID = ledgerId.bytes
            staking?.toProtobufInto(&proto.stakingInfo)
        }
    }
}
